### PR TITLE
feat(speechlm2): add generalized speech encoder foundations

### PR DESCRIPTION
**Pull request by pzelasko**
# feat(speechlm2): add generalized speech encoder foundations

> **Stack 1/7:** **Encoder foundations** → [2/7 Indexed/resumable data](https://github.com/NVIDIA-NeMo/Speech/pull/16193) → [3/7 Packed speech training](https://github.com/NVIDIA-NeMo/Speech/pull/16194) → [4/7 Portable export + target-only vLLM](https://github.com/NVIDIA-NeMo/Speech/pull/16195) → [5/7 MTP serving](https://github.com/NVIDIA-NeMo/Speech/pull/16196) → [6/7 DFlash serving](https://github.com/NVIDIA-NeMo/Speech/pull/16239) → [7/7 CCE dependency](https://github.com/NVIDIA-NeMo/Speech/pull/16241)

Base commit (`main` at rebase): `abb8254dac2bf5a011e6069fcaa7df71c7e3b8c1`

Head: `codex/pr16158-rewrite/01-encoder-foundations`

Tip: `1eae02f8b05f8a6ca00fa7320546c62d6de983bb`

## Summary

- Add Parallel Expert Encoder composition and resolution over the maintained Transformer/Sortformer components.
- Add bounded encoder chunk microbatching with synchronized FSDP forward counts.
- Enforce exact chunk/stride compatibility and retain the audio-grid cap used by the 30-second dual-encoder path.
- Define portable speaker-feature/alignment contracts and the Nemotron 3.5 prompt schema.
- Add process-group timeout propagation shared by dependent training paths.
- Remove the unused GGEMM/MoE encoder prototypes, their exports, and their tests.
- Remove obsolete SALM-only `no_rttm_to_ones` and `phPEE` terminology.

## Reviewer guide

The review question is: are the reusable encoder, speaker, chunking, and prompt contracts correct before indexed data, padding-free execution, export, or vLLM are introduced?

Native THD execution, indexed data, checkpoint publication, and serving are intentionally absent.

## Attribution

The reconstructed commits retain their original authors, dates, source-commit references, and author-matching DCO sign-offs. Taejin Park's substantive review fix is preserved as its own authored commit; redundant merge/rebase copies are omitted.

## Validation

- `git diff --check`: passed.
- Branch-scoped pre-commit: passed.
- Compilation of every changed Python file: passed.
- Focused stack suite: 403 passed, 20 skipped.
- Chunking tests cover the initialized-distributed synchronization path, zero/gradient-neutral dummy loss, stride validation, and the retained audio-grid cap.
- The `StreamingTransformerEncoder` docstring parses without the inline-markup warning that failed Sphinx CI.
- All stack commits have author-matching DCO sign-offs.

### CODE REVIEW

File: nemo/collections/asr/modules/parallel_expert_encoder_ggemm.py [hunk 1]
```diff
@@ -0,0 +1,1938 @@
...
+        return audio_signal, length.to(device=audio_signal.device)
+
+    # Forward — identical signature to ConformerEncoder.forward
+    def forward(
+        self,
+        audio_signal,
+        length,
+        spk_targets=None,
+        return_experts: bool = False,
+    ):
```
[the remark this answers is not in the document]
Comment (pzelasko) [reply nemo/collections/asr/modules/parallel_expert_encoder_ggemm.py]: This is intentional and documented: the default two-value return preserves the `ConformerEncoder.forward` contract, while `return_experts=True` is an explicit diagnostic opt-in that adds the expert outputs. Making the default arity three would break existing encoder callers, so no code change is appropriate here.

File: nemo/collections/asr/modules/parallel_expert_encoder_two_branch.py [hunk 2]
```diff
@@ -0,0 +1,31 @@
...
+"""
+
+from nemo.collections.asr.modules.parallel_expert_encoder import *  # noqa: F401,F403
+from nemo.collections.asr.modules.parallel_expert_encoder import (  # noqa: F401
+    ParallelExpertEncoder,
+    ParallelExpertEncoderPT,
+    _clone_config,
+    _default_dtype,
+    _disable_dist_feature_sync,
+)
```
[the remark this answers is not in the document]
Comment (pzelasko) [reply nemo/collections/asr/modules/parallel_expert_encoder_two_branch.py]: Resolved in `2194eadf4bad9742702edd364d2b2edcd0badaa2`: the unused compatibility imports were removed from this PR, and linting/CodeQL now pass at the current head.

File: nemo/collections/asr/modules/moe_transformer_encoder.py
Comment (nithinraok) [inline nemo/collections/asr/modules/moe_transformer_encoder.py:1]: Do we need this now? We are not using moe expert encoder for the current model.

File: nemo/collections/asr/modules/ggemm_transformer_encoder.py
Comment (nithinraok) [inline nemo/collections/asr/modules/ggemm_transformer_encoder.py:1]: Same, Do we need this now? We are not using this version of expert encoder for the current model, correct?

Comment (nithinraok) [reply nemo/collections/asr/modules/ggemm_transformer_encoder.py:1]: And no test coverage for this class too

File: nemo/collections/asr/parts/utils/sot_speaker_alignment.py [hunk 3]
```diff
@@ -25,13 +25,31 @@
 SPEAKER_TOKEN_PATTERN = re.compile(r"<spk:(\d+)>")
 _SPEAKER_TOKEN_SPLIT_PATTERN = re.compile(r"(<spk:\d+>)")
 
+# SOT speaker alignment is only used to resolve the RTTM column permutation; the
+# full-resolution activity tensor is returned unchanged apart from that column
+# reorder. Bounding the DTW input to 1,200 frames keeps its
+# O(words * frames * permutations) cost stable for long-form sessions. The
+# effective frame duration is max(80 ms, utterance_duration / 1,200): short inputs
+# are never upsampled, and every coarse bin consumes at least one real input frame.
+# A one-hour session therefore uses 1,200 bins of 37 or 38 frames, or 3.0 seconds each.
```
Comment (nithinraok) [inline nemo/collections/asr/parts/utils/sot_speaker_alignment.py]: Reduce comment size

File: tests/collections/speechlm2/test_encoder_chunking.py
Comment (nithinraok) [inline tests/collections/speechlm2/test_encoder_chunking.py:1]: We are only checking with syncing off when chunking is enabled. Shouldn't we test with `dist.is_initialized` as `True` ?

File: nemo/collections/speechlm2/parts/encoder_chunking.py [hunk 4]
```diff
@@ -46,30 +52,53 @@ def encode_audio_with_optional_chunking(
...
+            input_signal,
+            input_signal_length,
+            chunk_size_seconds=chunk_size_seconds,
+            sampling_rate=sampling_rate,
+            chunk_batch_size=chunk_batch_size,
+            sync_group=sync_group,
+        )
+        return _maybe_return_dummy_loss([], dummy_loss, return_dummy_loss)
 
     chunk_size_samples = _get_chunk_size_samples(chunk_size_seconds, sampling_rate)
```
Comment (nithinraok) [inline nemo/collections/speechlm2/parts/encoder_chunking.py:86]: are we ensuring `chunk_size_samples` is multiple of `stride`

File: tests/collections/speechlm2/test_encoder_chunking.py [hunk 5]
```diff
@@ -182,6 +338,30 @@ def test_encode_audio_with_optional_chunking_does_not_forward_absent_spk_targets
...
+        perception,
+        audios,
+        audio_lens,
+        chunk_size_seconds=1.0,
+        sampling_rate=2,
+        chunk_batch_size=2,
+        return_dummy_loss=True,
+    )
+
+    assert dummy_loss is None
```
Comment (nithinraok) [inline tests/collections/speechlm2/test_encoder_chunking.py:374]: shouldn;t we test for `dummy_loss` to zero? 

Also, should we test if adding this `dummy_loss` leaves gradients unchanged.

File: nemo/collections/speechlm2/models/salm_automodel.py [hunk 6]
```diff
@@ -242,56 +246,57 @@ def forward(
...
+            yield
             return
+        with self.perception.encoder.online_inference():
+            yield
 
-        unsupported = []
+    def _warn_parallel_expert_encoder_inference_chunking(self) -> None:
+        if not self.cfg.get("pe_encoder_path", None):
+            return
         if self.cfg.get("encoder_chunk_size_seconds", None) is not None:
```
Comment (nithinraok) [inline nemo/collections/speechlm2/models/salm_automodel.py:274]: making sure if we are using `40s` or something different.

File: nemo/collections/asr/parts/utils/sot_speaker_alignment.py [hunk 7]
```diff
@@ -316,7 +544,9 @@ def speaker_activity_from_cut(
     num_mel_frame_per_target_frame: int,
     no_rttm_to_ones: bool = True,
```
Comment (nithinraok) [inline nemo/collections/asr/parts/utils/sot_speaker_alignment.py:540]: in `_build_speaker_activities` this is being overridden with `torch.full_like(..., -1.0),` so we are not using this flag

File: nemo/collections/asr/modules/parallel_expert_encoder.py [hunk 8]
```diff
@@ -39,87 +43,249 @@
...
+            diar_spkcache_len=self._cfg.get("diar_spkcache_len", 188),
+            missing_rttm_target=self._cfg.get("missing_rttm_target", -1.0),
+            speaker_feature_mode=speaker_feature_mode,
+            speaker_activity_threshold=speaker_activity_threshold,
+            spk_kernel_scale=self._cfg.get("spk_kernel_scale", 1.0),
+            sync_max_audio_length=self._cfg.get("sync_max_audio_length", False),
         )
+        # Keep the architecture-only bundle config beside the inner module.
+        # SpeechLM HF export embeds this small config in config.json so the
+        # consolidated checkpoint can reconstruct phPEE without carrying a
```
Comment (nithinraok) [inline nemo/collections/asr/modules/parallel_expert_encoder.py]: remove `phPEE` references

**Review by nithinraok — CHANGES_REQUESTED** [review]

File: nemo/collections/speechlm2/models/salm_automodel.py (hunk 6 above)
Comment (pzelasko) [reply nemo/collections/speechlm2/models/salm_automodel.py:274]: we're using 30s

File: nemo/collections/asr/parts/utils/sot_speaker_alignment.py (hunk 7 above)
Comment (tango4j) [reply nemo/collections/asr/parts/utils/sot_speaker_alignment.py:540]: That overriding only applies to salm_automodel.py and salm_dataset.py. 
The code review agent only confined the scope to salm_automodel.py.
`sot_speaker_alignment.py` is also used for multi-talker ASR models, and `no_rttm_to_ones` option is not override for other use cases. 

`no_rttm_to_ones` are all removed from `salm_dataset.py` for this reason.

File: nemo/collections/speechlm2/parts/encoder_chunking.py (hunk 4 above)
Comment (tango4j) [reply nemo/collections/speechlm2/parts/encoder_chunking.py:86]: @pzelasko 
This is out of my changes to the encoder_chunkking.py, but I am adding a validation lines for this.
(Surely, users can provide some illegal chunk sizes)

File: tests/collections/speechlm2/test_encoder_chunking.py (hunk 5 above)
Comment (tango4j) [reply tests/collections/speechlm2/test_encoder_chunking.py:374]: I did not implement this part, but I think `dummy_loss` should be forced to `None` because you cannot perform `backward()` on `None` thus there is no chance that gradients get updated.

File: tests/collections/speechlm2/test_encoder_chunking.py
Comment (tango4j) [reply tests/collections/speechlm2/test_encoder_chunking.py:1]: The current chunk-microbatch test does not provide a synchronization group, so that is why this test only exercises the unsynchronized path.

File: nemo/collections/asr/modules/parallel_expert_encoder.py (hunk 8 above)
Comment (tango4j) [reply nemo/collections/asr/modules/parallel_expert_encoder.py]: removed

File: tests/collections/speechlm2/test_encoder_chunking.py (hunk 5 above)
Comment (nithinraok) [reply tests/collections/speechlm2/test_encoder_chunking.py:374]: But we are not ensuring `dummy_loss` to None always in training code?

**Review by nithinraok — DISMISSED** [review]
LGTM

File: tests/collections/speechlm2/test_encoder_chunking.py (hunk 5 above)
Comment (pzelasko) [reply tests/collections/speechlm2/test_encoder_chunking.py:374]: Yes—but that assertion covers the ordinary case where no dummy forward is needed.

In `test_encode_audio_with_optional_chunking_can_microbatch_chunks`:

- No `sync_group` is provided.
- There are two real microbatches.
- No rank-padding forward occurs.
- Therefore `dummy_loss is None` is correct.

The new test immediately below exercises the complementary case:

- A synchronization group is supplied.
- The mocked all-reduce raises the required count from one real microbatch to two.
- The second forward is a dummy.
- It asserts `dummy_loss is not None`, scalar, zero-valued, and gradient-neutral.

So the intended invariant is:

```python
dummy_loss is None       # no dummy forwards occurred
dummy_loss is zero tensor  # dummy forwards occurred and need an autograd edge
```

Comment (tango4j) [reply tests/collections/speechlm2/test_encoder_chunking.py:374]: https://github.com/NVIDIA-NeMo/Speech/blob/1ea4ec08610635470947bcff1424c5ef14db0776/nemo/collections/speechlm2/parts/encoder_chunking.py#L213

Whenever dummy_loss is not None, it is multiplied with 0.0.
So it is obvious that it does not produce any gradient,
but if @pzelasko agrees, we can put a simple verification line that this is true zero 0.0 somwhere 

But I personally feel like this is unnecessary check.

**Review by tango4j — APPROVED** [review]
LGTM
